### PR TITLE
 Implement debugger call stack view

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Screen Shot 2020-04-23 at 7 29 08 AM

### CODE REVIEW

File: packages/devtools_app/lib/src/debugger/flutter/debugger_model.dart [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  int get tokenPos => location.tokenPos;
+
+  @override
+  int get line => sourcePosition?.line ?? location.line;
+
+  @override
+  int get column => sourcePosition?.column ?? location.column;
+}
+
+/// A tuple of a stack frame and a source position.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_model.dart:126]: This is the only new code in this file. The rest are moves.

File: packages/devtools_app/lib/src/debugger/flutter/call_stack.dart [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import 'package:flutter/material.dart' hide Stack;
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/call_stack.dart:1]: nit: copyright

File: packages/devtools_app/lib/src/debugger/flutter/call_stack.dart [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      builder: (context, stackFrames, _) {
+        return ValueListenableBuilder<StackFrameAndSourcePosition>(
+          valueListenable: controller.selectedStackFrame,
+          builder: (context, selectedFrame, _) {
+            return ListView.builder(
+              itemCount: stackFrames.length,
+              itemExtent: defaultListItemHeight,
+              itemBuilder: (_, index) {
+                final frame = stackFrames[index];
+                final selected =
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/call_stack.dart:41]: You may want to just have `final selected = frame == selectedFrame`, and assume that code outside of here will select the first frame when it should.

File: packages/devtools_app/lib/src/debugger/flutter/call_stack.dart [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    var name = frame.frame.code?.name ?? '<none>';
+    if (name.startsWith('[Unoptimized] ')) {
+      name = name.substring('[Unoptimized] '.length);
+    }
+    name = name.replaceAll('<anonymous closure>', '<closure>');
+
+    if (frame.frame.kind == FrameKind.kAsyncSuspensionMarker) {
+      name = '<async break>';
+    }
+    return name;
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/call_stack.dart:111]: Perhaps have the name end with `()`?

Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/call_stack.dart:111]: (not for async markers, but for executable frames)

File: packages/devtools_app/lib/src/debugger/flutter/call_stack.dart [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      await controller.selectScript(frame.script);
+    } else if (frame.scriptUri != null) {
+      await controller
+          .selectScript(controller.scriptRefForUri(frame.scriptUri));
+    }
+  }
+
+  String _descriptionFor(StackFrameAndSourcePosition frame) {
+    var name = frame.frame.code?.name ?? '<none>';
+    if (name.startsWith('[Unoptimized] ')) {
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/call_stack.dart:103]: use a `const` here -

File: packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart [hunk 6]
```diff
@@ -297,13 +308,18 @@ class DebuggerController extends DisposableController
...
+      _stackFramesWithLocation.value = list.toList();
+    });
 
-    if (_currentStack.value != null && _currentStack.value.frames.isNotEmpty) {
+    if (frames.isNotEmpty) {
       // TODO(https://github.com/flutter/devtools/issues/1648): Allow choice of
       // the scripts on the stack.
-      _currentScript.value =
-          await getScript(_currentStack.value.frames.first.location.script);
+      _currentScript.value = await getScript(frames.first.location.script);
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart:322]: also set `_selectedStackFrame.value`?

File: packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart [hunk 7]
```diff
@@ -404,124 +420,55 @@ class DebuggerController extends DisposableController
...
+    if (_callStack.value == null) return [];
+
+    List<Frame> frames =
+        _callStack.value.asyncCausalFrames ?? _callStack.value.frames;
+
+    // Handle breaking-on-exceptions.
+    if (_reportedException != null && frames.isNotEmpty) {
+      final Frame frame = frames.first;
+
+      final Frame newFrame = Frame(
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart:451]: 👍 this is great! You're creating a synthetic frame to hold the exception value?

We could also consider adding a new, synthetic variable to the top frame, in the 1st position in the list of variables.

File: packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart [hunk 8]
```diff
@@ -404,124 +420,55 @@ class DebuggerController extends DisposableController
...
+        code: frame.code,
+        location: frame.location,
+        kind: frame.kind,
+      );
+
+      final List<BoundVariable> newVars = <BoundVariable>[];
+      newVars.add(BoundVariable(
+        name: '<exception>',
+        value: _reportedException,
+        scopeStartTokenPos: null,
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart:463]: Is it necessary to pass nulls here?

File: packages/devtools_app/test/flutter/debugger_screen_test.dart [hunk 9]
```diff
@@ -156,6 +161,125 @@ void main() {
       );
     });
 
+    testWidgetsWithWindowSize(
+        'call stack show items', const Size(1000.0, 4000.0),
```
Comment (devoncarew) [inline packages/devtools_app/test/flutter/debugger_screen_test.dart:165]: Interesting, I didn't know you could adjust the test device size.

File: packages/devtools_app/lib/src/debugger/flutter/call_stack.dart [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+          alignment: Alignment.centerLeft,
+          child: RichText(
+            maxLines: 1,
+            overflow: TextOverflow.ellipsis,
+            text: TextSpan(
+              text: _descriptionFor(frame),
+              style: selected ? selectedStyle : regularStyle,
+              children: [
+                TextSpan(
+                  text: ' (${frame.scriptUri}:${frame.line})',
```
Comment (devoncarew) [inline packages/devtools_app/lib/src/debugger/flutter/call_stack.dart:77]: The `line` info is going to be pushed far off to the right.

After this lands, we may want to play with different visualizations. `method() file:line` ? That's terse, and the script uri info would be available when they click on the frame.

File: packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart (hunk 7 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart]: Ah, I didn't read the code closely enough; this does create a synthetic var in the first frame.

**Pull request merged**